beacon_node/beacon_chain/src/sync_committee_verification.rs

```diff
@@ -0,0 +1,713 @@
+        //
+        // It's important to double check that the contribution is not already known, otherwise two
+        // contribution processed at the same time could be published.
+        if let ObserveOutcome::AlreadyKnown = chain
+            .observed_sync_contributions
+            .write()
+            .observe_item(contribution, Some(contribution_root))
+            .map_err(|e| Error::BeaconChainError(e.into()))?
+        {
+            return Err(Error::SyncContributionAlreadyKnown(contribution_root));
+        }
+
+        // Observe the aggregator so we don't process another aggregate from them.
+        //
+        // It's important to double check that the sync committee message is not already known, otherwise two
+        // sync committee messages processed at the same time could be published.
+        if chain
+            .observed_sync_aggregators
+            .write()
+            .observe_validator(observed_key, aggregator_index as usize)
+            .map_err(BeaconChainError::from)?
+        {
+            return Err(Error::PriorSyncCommitteeMessageKnown {
+                validator_index: aggregator_index,
+                slot: contribution.slot,
+            });
+        }
+        Ok(VerifiedSyncContribution { signed_aggregate })
+    }
+
+    /// A helper function to add this aggregate to `beacon_chain.op_pool`.
+    pub fn add_to_pool(self, chain: &BeaconChain<T>) -> Result<(), Error> {
+        chain.add_contribution_to_block_inclusion_pool(self)
+    }
+
+    /// Returns the underlying `contribution` for the `signed_aggregate`.
+    pub fn contribution(self) -> SyncCommitteeContribution<T::EthSpec> {
+        self.signed_aggregate.message.contribution
+    }
+
+    /// Returns the underlying `signed_aggregate`.
+    pub fn aggregate(&self) -> &SignedContributionAndProof<T::EthSpec> {
+        &self.signed_aggregate
+    }
+}
+
+impl VerifiedSyncCommitteeMessage {
+    /// Returns `Ok(Self)` if the `sync_message` is valid to be (re)published on the gossip
+    /// network.
+    ///
+    /// `subnet_id` is the subnet from which we received this sync message. This function will
+    /// verify that it was received on the correct subnet.
+    pub fn verify<T: BeaconChainTypes>(
+        sync_message: SyncCommitteeMessage,
+        subnet_id: SyncSubnetId,
+        chain: &BeaconChain<T>,
+    ) -> Result<Self, Error> {
+        // Ensure sync committee message is for the current slot (within a
+        // MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance).
+        //
+        // We do not queue future sync committee messages for later processing.
+        verify_propagation_slot_range(chain, &sync_message)?;
+
+        // Sync messages must be for a known block. If the block is unknown, we simply drop the
+        // sync committee message and do not delay consideration for later.
+        verify_head_block_is_known(chain, sync_message.beacon_block_root)?;
```

Comment:
We also need to remove this check, and move the check on `observed_sync_contributors` to after the subnet position check (which also guarantees membership in the sync committee).

In terms of size bounds I think we can reduce `SyncContributorSlotHashSet::default_capacity` to `SYNC_COMMITTEE_SIZE // SYNC_COMMITTEE_SUBNET_COUNT` per the spec.